### PR TITLE
KMP Algorithm for Javascript

### PR DESCRIPTION
**Pull request by ritvij14**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?
Yes
(Write your answer here.)

### Description
This is the KMP Algorithm for pattern searching in Strings which is more efficient than the Naive search algorithm. It has the Time complexity of O(m+n) and space complexity of O(n).

VS Code output
image

(Write your answer here.)

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests
Issue #427 
(Write your answer here.)

### CODE REVIEW

**Comment by ritvij14** [comment]
@siddharth25pandey @Kajol-Kumari @atarax665 @iamrajiv @plazzy99 please review this. Also, I was confused as to how should I take input so I have worked with a sample case as I saw in other javascript programs.

File: JavaScript/search/KMPalgorithm.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+} else {
+  console.log(`Pattern is present in text at index ${result}`);
+}
+
+/*
+Time complexity = O(m+n)
+Space complexity = O(n)
+m = number of characters in text
+n = number of characters in pattern
+*/
```
Comment (Kajol-Kumari) [inline JavaScript/search/KMPalgorithm.js]: Please add one extra like at the end of the file.

**Review by Kajol-Kumari — CHANGES_REQUESTED** [review]
@ritvij14 please add one sample input/output in the commented section at the end of the file. Rest looks good to me :)

File: JavaScript/search/KMPalgorithm.js (hunk 1 above)
Comment (ritvij14) [reply JavaScript/search/KMPalgorithm.js]: @Kajol-Kumari I have added the output for the input I used. Please have a look.

File: JavaScript/search/KMPalgorithm.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      patIndex = 0;
+      textIndex++;
+    }
+  }
+
+  return -1;
+}
+
+result = kmpSearch('hakincodes', 'kin');
+if(result === -1) {
```
Comment (Kajol-Kumari) [inline JavaScript/search/KMPalgorithm.js:58]: Let's not hardcode the `input string` and `search string`. Please take this as input from the user.

**Review by Kajol-Kumari — CHANGES_REQUESTED** [review]

File: JavaScript/search/KMPalgorithm.js (hunk 2 above)
Comment (ritvij14) [reply JavaScript/search/KMPalgorithm.js:58]: @Kajol-Kumari but what should I use to take input? In the other Javascript codes in this repository, they were all using hardcoded inputs.

**Review by Kajol-Kumari — APPROVED** [review]
LGTM :+1:

**Review by vatsalkesarwani12 — APPROVED** [review]
Code works well 👍

**Review by siddharth25pandey — APPROVED** [review]
@ritvij14 Good to go
Able to merge.
Thanks!

**Pull request merged**